File: e-kotlin/src/main/kotlin/e/kotlin/Extensions.kt
package e.kotlin

/**
 * Computes a new EOr using E in this, if it exists, with given flat mapping function
 *
 * @param A Type of value
 *
 * @param f E flat mapping function
 *
 * @return This EOr or a computed EOr if this one has E
 */
inline fun <reified A> EOr<A>.flatMapError(crossinline f: (E) -> EOr<A>): EOr<A> =
    fold({ e -> f(e) }, { this })

/**
 * Gets the value in this or falls back to given default value
 *
 * @param A Type of value
 *
 * @param default Default value to use in case this has E
 *
 * @return Value in this or given default value
 */
inline fun <reified A> EOr<A>.getOrElse(crossinline default: () -> A): A =
    fold({ default() }, { a -> a })

/**
 * Provides an alternative EOr if this one has E, ignoring the E
 *
 * @param A Type of value
 *
 * @param alternative Alternative EOr in case this one has E
 *
 * @return This EOr or alternative if this one has E
 */
inline fun <reified A> EOr<A>.orElse(crossinline alternative: () -> EOr<A>): EOr<A> =
    fold({ alternative() }, { this })

/**
 * Converts this value to a successful EOr
 *
 * @return An EOr containing this value
 *
 * @see e.kotlin.EOr
 */
fun <A> A.toEOr(): EOr<A> =
    EOr.from(this)

/**
 * Constructs an EOr from this nullable value
 *
 * @param A Type of value
 *
 * @param ifNull An E to use in case this value is null
 *
 * @return An EOr containing either this value or given E
 */
fun <A> A?.toEOr(ifNull: () -> E): EOr<A> =
    EOr.fromNullable(this, ifNull)

/**
 * Constructs an EOr from this computation that can throw
 *
 * @param A Type of value
 *
 * @param ifFailure An E conversion function
 *
 * @return An EOr containing either computed value or an E computed by given function
 */
fun <A> (() -> A).catching(ifFailure: (Throwable) -> E = { t -> E.fromThrowable(t) }): EOr<A> =
    EOr.catching(this, ifFailure)

/**
 * Constructs an E from this [[java.lang.Throwable]]
 *
 * @return A new E containing message of this Throwable
 */
fun Throwable.toE(f: (Throwable) -> E = { t -> E.fromThrowable(t) }): E =
    f(this)

File: e-scala/src/main/scala/e/scala/extensions.scala
package e.scala

import _root_.scala.util.Try

/** Type alias for EOr, allowing a more pleasing syntax as following:
  *
  * {{{
  *   E or String // instead of EOr[String]
  * }}}
  *
  * @tparam A
  *   Type of the value EOr can contain
  * @see
  *   [[e.scala.E]]
  * @see
  *   [[e.scala.EOr]]
  */
infix type or[+_ <: E, +A] = EOr[A]

extension [A](a: => A)
    /** Converts this value to a successful EOr[A]
      *
      * @return
      *   An EOr[A] containing this value
      * @see
      *   [[e.scala.EOr]]
      */
    inline def toEOr: EOr[A] =
        EOr(a)

    /** Constructs an EOr from evaluating this value and by converting caught Exception using given function
      *
      * @param ifFailure
      *   An E conversion function in case evaluating this value throws
      * @return
      *   An EOr containing either evaluated value or an E computed by given function
      */
    inline def catching(ifFailure: Throwable => E = E.fromThrowable): EOr[A] =
        EOr.fromTry(Try(a))(ifFailure)

extension [A](option: Option[A])
    /** Constructs an EOr from this [[_root_.scala.Option]]
      *
      * @param ifNone
      *   An E to use in case this Option is None
      * @return
      *   An EOr containing either value in this Option or given E
      */
    inline def toEOr(ifNone: => E): EOr[A] =
        EOr.fromOption(option)(ifNone)

extension [L, R](either: Either[L, R])
    /** Constructs an EOr from this [[_root_.scala.util.Either]]
      *
      * @param ifLeft
      *   An E conversion function in case this Either is Left
      * @return
      *   An EOr containing either Right value in this Either or an E computed by given function
      */
    inline def toEOr(ifLeft: L => E): EOr[R] =
        EOr.fromEither(either)(ifLeft)

extension [A](t: Try[A])
    /** Constructs an EOr from this [[_root_.scala.util.Try]]
      *
      * @param ifFailure
      *   An E conversion function in case this Try is Failure
      * @return
      *   An EOr containing either value in this Try or an E computed by given function
      */
    inline def toEOr(ifFailure: Throwable => E = E.fromThrowable): EOr[A] =
        EOr.fromTry(t)(ifFailure)

extension (throwable: Throwable)
    /** Constructs an E from this [[_root_.scala.Throwable]]
      *
      * @param f
      *   A mapping function
      * @return
      *   A new E containing message of this Throwable
      */
    inline def toE(f: Throwable => E = E.fromThrowable): E =
        f(throwable)
